### PR TITLE
Version 0.7

### PR DESCRIPTION
**Pull request by stephenbach**
A PR for collecting sub PRs, discussion, etc. (not ready for review yet)

Goals for this version:
Pytorch classifiers
New installation model that uses conda for everything and is installable from pip
Switch to Spacy parser as default and support Spacy 2

### CODE REVIEW

**Comment by danich1** [comment]
Bump! What is the status on this pr? is there anything I can do to help get this merged?

File: snorkel/learning/pytorch/noise_aware_model.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        total_loss += target[:, i].dot(loss)
+    return total_loss / input.shape[0]
+
+
+class TorchNoiseAwareModel(Classifier, nn.Module):
+    """
+    Generic NoiseAwareModel class for PyTorch models.
+    
+    :param n_threads: Parallelism to use; single-threaded if None
+    :param seed: Top level seed which is passed into both numpy operations
```
Comment (ajratner) [inline snorkel/learning/pytorch/noise_aware_model.py]: This doesn't seem to actually be handled anywhere?  Also this is a fun place of trickiness... need to make sure to seed any other random modules (`random, numpy.random`) as well as potentially give options to turn off CuDNN for determinism, e.g. see https://github.com/HazyResearch/metal_nips/blob/666eea29ecbc9f816e3e9ae7690200409610ff09/metal/metal_model.py#L205

File: snorkel/learning/pytorch/noise_aware_model.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    def _build_model(self, **model_kwargs):
+        raise NotImplementedError
+    
+    def marginals(self, X, batch_size=None, **kwargs):
+        nn.Module.train(self, False)
+        marginals = self._pytorch_outputs(X, batch_size).detach()
+        return F.sigmoid(marginals).numpy() if self.cardinality == 2 else F.softmax(marginals).numpy()
+
+    def _pytorch_outputs(self, X, batch_size):
```
Comment (ajratner) [inline snorkel/learning/pytorch/noise_aware_model.py]: Comments + docstring here and the method above :)

File: snorkel/learning/pytorch/noise_aware_model.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        :param kwargs: All hyperparameters that change how the network is built
+            must be passed through here to be saved and reloaded to save /
+            reload model. *NOTE: If a parameter needed to build the 
+            network and/or is needed at test time is not included here, the
+            model will not be able to be reloaded!*
+        """
+        # Sets random seeds
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed(seed)
+        torch.cuda.manual_seed_all(seed)
```
Comment (ajratner) [inline snorkel/learning/pytorch/noise_aware_model.py]: See comment above- might want to add in option to set CuDNN off in case that causes non-determinism?

File: snorkel/learning/pytorch/logistic_regression.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        self.linear = nn.Linear(input_size, num_classes)
+        self.sigmoid = nn.Sigmoid()
+
+    def forward(self, X):
+        out = self.linear(X)
+        return self.sigmoid(out)
+        
+    def _check_input(self, X):
+        if issparse(X):
+            raise Exception("Sparse input matrix. Use SparseLogisticRegression")
```
Comment (ajratner) [inline snorkel/learning/pytorch/logistic_regression.py]: Since there's no PyTorch version of `SparseLogisticRegression` implemented here yet, just cast to dense and print a warning about this?

**Comment by dhimmel** [comment]
Thanks @ajratner for the review, as hopefully this can get merged soon.

One issue with the current approach is that the initial pull requests with changes have already been merged into this massive PR. Therefore, it's not clear whose responsibility it is to make the requested changes. Do you want @danich1, @stephenbach, and I to make the changes on the parts we each originally contributed? To do that, we will need to open additional pull requests into the `0.7.0-alpha` branch, which can then be merged such that this PR then addresses your requested changes above.

I do not have much time to continue working on follow ups related to #921. Ideally, review would have taken place on that pull request, so once it was merged my involvement was complete. However, I am willing to do what it takes to see the work from #921 make it into `master`, and preferably without much more delay. So please tag me on the individual comments above if you'd like me to address them, and I will do so.

**Comment by stephenbach** [comment]
@dhimmel Sorry, I'm the bottleneck now. I'll make the changes, hopefully this weekend, and merge them in. :)

**Comment by dhimmel** [comment]
> I'll make the changes, hopefully this weekend

Thanks for the update. Let us know if you'd like any help.

File: README.md [hunk 5]
```diff
@@ -48,74 +47,141 @@ However, **_Snorkel is very much a work in progress_**, so we're eager for any a
...
+In addition, we recommend installing packages via pip, only if they are not available in the conda-forge channel.
+3. Add the `snorkel` package installation to your `environment.yml`, under the `- pip` section.
+Of course, we suggest versioning snorkel, which you can do via a release number or commit hash (to access more bleeding edge functionality)
+  ```yml
+    # Versioned via release tag
+    - git+https://github.com/HazyResearch/snorkel@v0.6.3
+    # Versioned via commit hash (commit hash below is fake to ensure you change it)
+    - git+https://github.com/HazyResearch/snorkel@7eb7076f70078c06bef9752f22acf92fd86e616a
+  ```
+Finally, consider versioning the `numbskull` and `treedlib` pip dependencies by changing `master` to their latest commit hash on GitHub.
```
[the remark this answers is not in the document]
Comment (stephenbach) [reply README.md]: I wouldn't object, but might be good to keep support for some automatic featurization. It's essentially required for using logistic regression. At least now it's a separate package, not a git submodule, so should be invisible to people who don't use it.

File: snorkel/learning/pytorch/noise_aware_model.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+from snorkel.learning.classifier import Classifier
+from snorkel.learning.utils import reshape_marginals, LabelBalancer
+
+
+def cross_entropy_loss(input, target):
+    total_loss = torch.tensor(0.0)
+    for i in range(input.size(1)):
+        cls_idx = torch.full((input.size(0),), i, dtype=torch.long)
+        loss = F.cross_entropy(input, cls_idx, reduce=False)
```
[the remark this answers is not in the document]
Comment (stephenbach) [reply snorkel/learning/pytorch/noise_aware_model.py]: Sorry, I don't follow. Just check that input and target have the same dimensions and dim 0 is example and dim 1 is class?

File: snorkel/learning/pytorch/logistic_regression.py (hunk 4 above)
Comment (stephenbach) [reply snorkel/learning/pytorch/logistic_regression.py]: Done 75dc24a

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (ajratner) [reply README.md]: Okay that's fine!

File: README.md [hunk 7]
```diff
@@ -48,74 +47,141 @@ However, **_Snorkel is very much a work in progress_**, so we're eager for any a
...
+We recommend starting with the [`environment.yml`](environment.yml) in this repository.
+The below modifications can help customize it for your needs:
+
+<details>
+
+1. Specifying versions for the listed packages, such as changing `python` to `python=3.6.5`.
+Versioned specification of your environment is critical to reproducibility and ensuring dependency updates do not break your pipeline.
+When first setting your package versions, you likely want to start with the latest versions available on the [conda-forge](https://anaconda.org/conda-forge/) channel, unless you have a reason to do otherwise.
+2. Adding other packages to your environment as required by your use case.
+Consider maintaining alphabetical sorting of packages in `environment.yml` to assist with maintainability.
```
[the remark this answers is not in the document]
Comment (ajratner) [reply README.md]: I think this is fine to start- main thing is the change to "Quick start" (small but important IMO)- thanks for implementing!

File: snorkel/learning/pytorch/noise_aware_model.py (hunk 3 above)
Comment (ajratner) [reply snorkel/learning/pytorch/noise_aware_model.py]: Yeah I agree, good point

File: snorkel/learning/pytorch/noise_aware_model.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/learning/pytorch/noise_aware_model.py]: nvm!
